sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala

```diff
@@ -236,15 +241,22 @@ private[sql] object PartitioningUtils {
   }
 
   /**
-   * Converts a string to a [[Literal]] with automatic type inference.  Currently only supports
-   * [[IntegerType]], [[LongType]], [[DoubleType]], [[DecimalType.SYSTEM_DEFAULT]], and
-   * [[StringType]].
+   * Converts a string to a [[Literal]] with automatic type inference if no field type specified.
+   * Auto inference only supports [[IntegerType]], [[LongType]], [[DoubleType]],
+   * [[DecimalType.SYSTEM_DEFAULT]], and [[StringType]].
    */
   private[sql] def inferPartitionColumnValue(
+      expectedDT: Option[DataType],
```

Comment:
We need to pass the expect the data type down and then get the associated literal-based partition column value; and @liancheng's suggestion kind of like get the literal (maybe string based) first, and then do casting outside, however, this probably lose some data precision during the re-casting.

For example:
The path looks like, /part1=1.000, and with the auto inference, we will get a Double, and it will be cast to string as `1.0` if what user expect is StringType;

However, this is totally different if we get it as StringType directly, which supposed to be `1.000`.